### PR TITLE
Consolidate NOS Standard Suite metrics into single source-of-truth module

### PR DESCRIPTION
**Pull request by kammereraj**
### Description of Changes

This PR consolidates the NOS Standard Suite metrics (RMSE, CF, POF, NOF, bias, stdev, Pearson R) into a single source-of-truth module and fixes several inconsistencies across the codebase. Previously, these metrics were computed inline in 3 separate files with slight differences.

This PR also extends `nos_metrics` adoption to the **ice skill assessment** code. The ice files (`find_ofs_ice_stations.py`, `do_iceskill.py`) computed RMSE, Pearson R, bias, and stdev inline using raw numpy/scipy calls. These have been replaced with `nos_metrics` calls where applicable, and `nos_metrics.rmse` has been made NaN-safe to support ice data (which contains NaN land masks).

**New file:**
- `src/ofs_skill/skill_assessment/nos_metrics.py` — Pure-function module containing all NOS Standard Suite metric calculations. No DataFrame coupling, no rounding, no logger, no side effects.

**New metrics (MDPO/MDNO):**
- `max_duration_positive_outliers(errors, threshold)` — Longest consecutive run of errors >= 2*threshold (NaN breaks streak)
- `max_duration_negative_outliers(errors, threshold)` — Longest consecutive run of errors <= -2*threshold (NaN breaks streak)
- These were missing from the NOS Standard Suite implementation and are now available for use.

**Ice metrics consolidation:**
- `nos_metrics.rmse` switched from sklearn's `root_mean_squared_error` (not NaN-safe) to `np.sqrt(np.nanmean(...))` (NaN-safe). Results are identical for NaN-free data; now also handles ice arrays with NaN land masks. The `sklearn.metrics` import is removed from `nos_metrics.py`.
- `find_ofs_ice_stations.py` — replaced 4 inline calculations (Pearson R via `scipy.stats.pearsonr`, RMSE, mean bias, stdev) with `nos_metrics` calls. Removed `from scipy import stats`.
- `do_iceskill.py` — replaced 2 inline RMSE calculations (`rmse_all`, `rmse_either`) with `nos_metrics.rmse()`.
- **Not replaced** (intentionally left as-is): 2D RMSE with `axis=0` (returns spatial arrays), skill score, CSI/confusion matrix, mean ice cover calculations, and both RMSE calls in `make_ice_boxplots.py` (also `axis=0`).

**Bug fixes:**
1. **`<` vs `<=` in `skill_scalar()` CF calculation** — `metrics_paired_one_d.py` `skill_scalar()` used strict `<` for central frequency (`-X1 < npbias & npbias < X1`), while `skill_vector()` and all 2D code used `<=`. NOS convention is `<=` (inclusive). Now all callers use `nos_metrics.central_frequency()` which uses `<=`.
2. **Tab-separator bug in `make_skill_maps.get_error_range()`** — This copy wrote default CSV with `sep='\t'` instead of comma-separated, which would produce an unreadable file on first run. Deleted entirely.
3. **Hardcoded `errorrange = 3` in `metrics_two_d.py`** — Both `return_one_d()` and `return_two_d()` now accept an optional `errorrange` parameter (default=3) for configurability while remaining backward-compatible.

**Files modified:**
| File | Change |
|------|--------|
| `src/ofs_skill/skill_assessment/nos_metrics.py` | **NEW** — Single source of truth for all metric functions; RMSE uses NaN-safe numpy (no sklearn dependency) |
| `src/ofs_skill/skill_assessment/metrics_paired_one_d.py` | Deleted `get_error_range()`, replaced inline metrics with `nos_metrics.*` calls, removed unused `sklearn`/`scipy` imports |
| `src/ofs_skill/skill_assessment/metrics_two_d.py` | Replaced inline CF/POF/NOF with `nos_metrics.*` calls, added optional `errorrange` param |
| `src/ofs_skill/skill_assessment/make_skill_maps.py` | Deleted `get_error_range()` (had tab-separator bug, was never called internally) |
| `src/ofs_skill/visualization/plotting_functions.py` | Replaced `get_error_range()` body with thin wrapper delegating to `nos_metrics.get_error_threshold()`, preserving `(name_var, prop, logger)` signature |
| `src/ofs_skill/skill_assessment/__init__.py` | Removed `get_error_range` from `make_skill_maps` import, added all `nos_metrics` functions to package-level exports |
| `tests/nos_metrics_test.py` | **NEW** — 30 unit tests covering all functions in `nos_metrics.py` |
| `tests/test_package_imports.py` | Fixed `test_top_level_import` to validate semver format instead of hardcoded version; fixed `test_forecast_hours` import path after `get_forecast_hours.py` was folded into `do_horizon_skill_utils.py` |
| `tests/test_s3_fallback.py` | Fixed STOFS tests to use correct directory format (no `netcdf` subdirectory) |
| `tests/test_usgs_retrieval.py` | Fixed inventory tests to validate core columns compatible with installed version |
| `src/ofs_skill/model_processing/do_horizon_skill_utils.py` | Fixed `get_forecast_hours()` bug: `stofs_3d_atl`/`stofs_3d_pac` assigned forecast length to `fcstcycles` instead of `fcstlength` |
| `src/ofs_skill/obs_retrieval/find_ofs_ice_stations.py` | Replaced 4 inline metric calls with `nos_metrics.*`; removed `from scipy import stats` |
| `bin/skill_assessment/do_iceskill.py` | Replaced 2 inline RMSE calls with `nos_metrics.rmse()` |
| `tests/test_ice_nos_metrics.py` | **NEW** — 19 end-to-end ice tests verifying `nos_metrics` produces identical results to original inline formulas |

**Files NOT changed (verified safe):**
- `get_skill.py` — Calls `skill_scalar`/`skill_vector` by module ref; signatures unchanged
- `plotting_2d.py` — Calls `return_one_d`/`return_two_d` with 3 positional args; new `errorrange` param has default
- `plotting_scalar.py`, `plotting_vector.py`, `plotting_scalar_ice.py` — Import `get_error_range` from `plotting_functions`; wrapper preserves signature
- `plot_forecast_hours.py`, `make_static_plots.py`, `make_OM_view.py` — Call `plotting_functions.get_error_range()`; wrapper handles it
- `make_ice_boxplots.py` — Both RMSE calls use `axis=0` (return 2D spatial arrays); cannot use scalar `nos_metrics.rmse`
- `plotting_scalar_ice.py` — Already uses `get_error_range` wrapper; no changes needed

---

### Expected Outcome of Changes

- All NOS Standard Suite metrics — including those in the ice skill assessment — are now computed from a single canonical implementation, eliminating copy-paste drift.
- `nos_metrics.rmse` is now NaN-safe, supporting ice concentration arrays with NaN land masks while producing identical results for NaN-free 1D data.
- The CF `<=` boundary bug in `skill_scalar()` is fixed — scalar and vector metrics now produce consistent results following NOS convention.
- The `get_error_range()` tab-separator bug is eliminated — default CSV files will always be written correctly.
- MDPO/MDNO metrics are now available for future use.
- No public function signatures or return formats are changed. All existing callers (`get_skill.py`, plotting modules, `plotting_2d.py`) continue to work with zero modifications.
- Error thresholds are read correctly from CSV when available, with built-in defaults as fallback: `wl: (0.15, 0.5)`, `salt: (3.5, 0.5)`, `temp: (3.0, 0.5)`, `cu: (0.26, 0.5)`, `ice_conc: (10.0, 0.5)`.

---

### Developer Questions and Checklist

**Is this a high priority PR? If yes, why and is there a date by which it needs to be merged?**
No. This is a refactor and bug fix — no urgent deadline. Will need to be merged prior to harmoic analysis. 

**Are there any changes needed to how the code should be run for operational runs? (Commands, flags, job timings, etc.)**
No. All function signatures and return formats are preserved. The job runs identically.

**Does this update need to be be deployed for operational runs?**
Yes, a reinstall (`pip install -e .`) is needed to pick up the new `nos_metrics` module. No new dependencies are introduced — `scipy`, `numpy`, and `csv` (stdlib) are already in the environment. The `sklearn` import has been removed from `nos_metrics.py` (RMSE now uses pure numpy).

**Does this update require front end/web app changes? (If yes, provide documentation to Min)**
No.

**Does this impact code development work on adding SCHISM? (If yes, tag Atieh)**
No. The metric functions are OFS-agnostic.

**Does this impact the expected number of output files for integration tests (i.e., will the integration test fail even though code changes result in desired change in the number of outputs?)**
No. Output file count is unchanged.

- [x] Developer's name is removed throughout the code?
- [x] Did you add relevant labels to the PR?
- [x] Have you run pylint and is the code at an acceptable pylint score? (>8.0)
- [x] Jobs contain the appropriate file checking and handle errors for any missing data?
- [x] Is log free of critical ERRORs or WARNINGs? For errors that are not critical and can remain in code, is there sufficient handling and logging?

---

### Testing Instructions

**1. Unit tests (new module in isolation):**
```bash
pytest tests/test_nos_metrics.py -v
```
Expected: 30/30 pass. Covers RMSE (NaN-safe), Pearson R, CF (boundary `<=`), POF, NOF, MDPO, MDNO, check_nos_criteria, get_error_threshold (defaults, CSV read, unknown variable, fallback).

**2. Ice end-to-end tests (nos_metrics vs original inline formulas):**
```bash
pytest tests/test_ice_nos_metrics.py -v
```
Expected: 19/19 pass. Tests cover:
- **RMSE 1-D** — verifies `nos_metrics.rmse` matches `np.sqrt(np.nanmean((mod-obs)**2))` with NaN land masks, all-NaN arrays, single elements, mixed NaN positions
- **RMSE 2-D** — confirms `axis=0` spatial RMSE returns arrays (validating why it was NOT replaced)
- **Pearson R** — verifies `nos_metrics.pearson_r` matches `scipy.stats.pearsonr` for NaN-filtered ice data
- **Mean bias** — verifies `nos_metrics.mean_bias` matches `np.nanmean(errors)` including all-NaN edge case
- **Standard deviation** — verifies `nos_metrics.standard_deviation` matches `np.nanstd(errors)`
- **Rounding preservation** — confirms `np.round(nos_metrics.*(…), 2)` equals `np.round(<inline formula>, 2)` for all 4 metrics
- **Station pipeline** — reproduces the full `find_ofs_ice_stations.py` calculation sequence (NaN filtering, R, RMSE, bias, stdev) and verifies all 4 metrics match
- **Daily loop RMSE** — reproduces the `do_iceskill.py` time loop pattern for `rmse_all` and `rmse_either` with threshold masking

**3. Full test suite (regression):**
```bash
pytest tests/ -v
```
Expected: 98/98 pass. Includes fixes to pre-existing test failures in `test_s3_fallback.py` (STOFS directory format), `test_package_imports.py` (version check, import path), and `test_usgs_retrieval.py` (column assertions).

**4. Package import verification:**
```bash
python -c "from ofs_skill.skill_assessment import rmse, central_frequency, get_error_threshold; print(get_error_threshold('wl'))"
```
Expected: `(0.15, 0.5)`

**5. End-to-end create_1dplot run (wcofs nowcast, stations):**
```bash
python ./bin/visualization/create_1dplot.py \
  -o wcofs -p ./ \
  -s 2026-01-21T00:00:00Z -e 2026-01-22T00:00:00Z \
  -d MLLW -ws [nowcast] -t stations
```
Verify:
- `data/skill/stats/skill_wcofs_water_level_nowcast_stations.csv` — 10 stations, `target_error_range=0.15`, pass/fail values consistent with CF>=90/POF<=1/NOF<=1
- `data/skill/stats/skill_wcofs_water_temperature_nowcast_stations.csv` — 5 stations, `target_error_range=3.0`
- `data/visual/` — 10 water level timeseries HTML, 5 water temperature timeseries HTML, 2 RMSE map HTML (17 total)
- Log file contains no tracebacks related to `nos_metrics`, `metrics_paired_one_d`, `metrics_two_d`, `make_skill_maps`, or `plotting_functions`

**6. End-to-end ice run:**

---

### Reminder checklist for PR reviewer:
- [ ] Run PEP8 compliance tests to ensure the code follows best practices
- [ ] Check that documentation in the script is sufficient
- [ ] Validate the output values (if applicable) to make sure the calculations are correct and make scientific/physical sense; include a comment on the pull request including what validation you performed, for replicability.
- [ ] Test code on windows and linux, using several OFS as test cases
- [ ] When approving or commenting on a pull request, add information on the calls you use in case they need to be replicated.

### CODE REVIEW

**Review by plimbernoaa — CHANGES_REQUESTED** [review]
@kammereraj great work! it definitely makes sense to have all stats in one place, for dev purposes and for consistentcy across the code base. All six tests in the 'testing instructions' section passed. Calls are below. (I also found an unrelated bug in the horizon skill script during testing, and i will open an issue for that.)

Before merging, there is one item to clarify about CF/POF/NOF: the PR indicates that NOS convention is <=/>= (inclusive). The existing skill assessment tech reports are ambiguous about this, and the fortran code shows that it is exclusive (see screenshot below). Is there another source stating that NOS convention is inclusive? We should make sure we're all on the same page about this!

Image

Calls used:
`python ./bin/skill_assessment/do_iceskill.py -o lsofs -p ./ -s 2026-02-05T00:00:00Z -e 2026-02-12T00:00:00Z -ws nowcast -da True`
`python ./bin/visualization/create_1dplot.py -o cbofs -p ./ -s 2025-05-01T00:00:00Z -e 2025-05-02T00:00:00Z -so co-ops -d MLLW -ws forecast_b -hs`
`python ./bin/visualization/create_1dplot.py -o wcofs -p ./ -s 2026-01-21T00:00:00Z -e 2026-01-22T00:00:00Z -so co-ops -d MLLW -ws forecast_b -hs`
`python ./bin/visualization/create_1dplot.py -o wcofs -p ./ -s 2026-01-21T00:00:00Z -e 2026-01-22T00:00:00Z -so co-ops -d MLLW -ws nowcast`

**Comment by kammereraj** [comment]
> @kammereraj great work! it definitely makes sense to have all stats in one place, for dev purposes and for consistentcy across the code base. All six tests in the 'testing instructions' section passed. Calls are below. (I also found an unrelated bug in the horizon skill script during testing, and i will open an issue for that. EDIT: [issue created](https://github.com/NOAA-CO-OPS/dev-Next-Gen-NOS-OFS-Skill-Assessment/issues/34).)
> 
> Before merging, there is one item to clarify about CF/POF/NOF: the PR indicates that NOS convention is <=/>= (inclusive). The existing skill assessment tech reports are ambiguous about this, and the fortran code shows that it is exclusive (see screenshot below). Is there another source stating that NOS convention is inclusive? We should make sure we're all on the same page!
> 
> Image
> Calls used: `python ./bin/skill_assessment/do_iceskill.py -o lsofs -p ./ -s 2026-02-05T00:00:00Z -e 2026-02-12T00:00:00Z -ws nowcast -da True` `python ./bin/visualization/create_1dplot.py -o cbofs -p ./ -s 2025-05-01T00:00:00Z -e 2025-05-02T00:00:00Z -so co-ops -d MLLW -ws forecast_b -hs` `python ./bin/visualization/create_1dplot.py -o wcofs -p ./ -s 2026-01-21T00:00:00Z -e 2026-01-22T00:00:00Z -so co-ops -d MLLW -ws forecast_b -hs` `python ./bin/visualization/create_1dplot.py -o wcofs -p ./ -s 2026-01-21T00:00:00Z -e 2026-01-22T00:00:00Z -so co-ops -d MLLW -ws nowcast`

Thanks Pat! I checked the technical reports regarding this, and here's what I found:

Referencing: NOS CS 17 (Hess et al., 2003) and NOS CS 24 (Zhang et al., 2006)

  From Table 3 of both reports (NOS CS 17 defines it; NOS CS 24 reproduces it verbatim as "from Hess et al., 2003"):

  CF(X) — Central Frequency. Fraction (percentage) of errors that lie within the limits ±X.

  The complementary definitions are:
  POF(X) — Positive Outlier Frequency. Fraction (percentage) of errors that are greater than X.
  NOF(X) — Negative Outlier Frequency. Fraction (percentage) of errors that are less than −X.

  Why this proves <= (inclusive)

  The key is the logical complement between CF, POF, and NOF. The definitions form a partition of all errors:

  - CF counts errors that "lie within the limits ±X" (i.e., -X ≤ e ≤ X)
  - POF counts errors "greater than X" (i.e., e > X — strict)
  - NOF counts errors "less than −X" (i.e., e < −X — strict)

  For CF + POF + NOF to account for 100% of errors (which they must, by definition), the boundary values where e = ±X must
  be counted by CF (inclusive <=), not by POF/NOF (which use strict >/<). If CF used strict <, boundary values would not be counted.

   The phrase "lie within the limits" is the standard mathematical convention for a closed interval [-X, +X], meaning <=. Best guess, is maybe the implementation in [sorc/skills.f](https://github.com/NOAA-CO-OPS/NOS-OFS-Skill-Assessment-Code/blob/main/sorc/skills.f) was a mistake?

  Citations

  Primary (original definition):
  Hess, K.W., T.F. Gross, R.A. Schmalz, J.G.W. Kelley, F. Aikman III, E. Wei, and M.S. Vincent, 2003: NOS Standards for
  Evaluating Operational Nowcast and Forecast Hydrodynamic Model Systems. NOAA Technical Report NOS CS 17, 46 pp.
  - Table 3 (p. 9): Metric definitions
  - Equation (5) (p. 10): CF(X1) ≥ 90%, POF(X2) ≤ 1%, NOF(X2) ≤ 1%
  - PDF: https://tidesandcurrents.noaa.gov/ofs/publications/CS_Techrpt_017_SkillAss_Standards_2003.pdf

  Secondary (software implementation reference):
  Zhang, A., K.W. Hess, E. Wei, and E. Myers, 2006: Implementation of Model Skill Assessment Software for Water Level and
  Current in Tidal Regions. NOAA Technical Memorandum NOS CS 24, 61 pp.
  - Section 2.3 (p. 5): Reproduces Table 3 definitions "from Hess et al., 2003"
  - Table 4 (p. 6): Standard Suite criteria with CF(X) ≥ 90%
  - PDF: https://tidesandcurrents.noaa.gov/ofs/publications/CS_Techrpt_024_SkillAss_WLsCUs_2006.pdf

  The Bug Summary:

  The old skill_scalar() code used -X1 < npbias & npbias < X1 (strict <), which is inconsistent with the NOS definition.
  skill_vector() and the 2D code already correctly used <=. The consolidated nos_metrics.central_frequency() now uses <=
  everywhere, matching the NOS convention defined in Hess et al. (2003).
